p4runtime_sh/shell.py

```diff
@@ -2263,6 +2280,195 @@ def add(self, egress_port=None, instance=0):
+        md = p4runtime_pb2.PacketMetadata()
+        md.metadata_id = md_info.id
+        md.value = bytes_utils.parse_value(value.strip(), md_info.bitwidth)
+        return md
+
+    def __setitem__(self, name, value):
+        md_info = self._get_md_info(name)
+        self._md[name] = self._parse_md(value, md_info)
+
+    def _ipython_key_completions_(self):
+        return self._md_info.keys()
+
+    def set(self, **kwargs):
+        for name, value in kwargs.items():
+            self[name] = value
+
+    def clear(self):
+        self._md.clear()
+
+    def values(self):
+        return self._md.values()
+
+
+latest_hid = 0
+packet_in_handles = {}  # Handle ID -> handle function
+class PacketIn():
+
+    def __init__(self):
+        ctrl_pkt_md = P4Objects(P4Type.controller_packet_metadata)
+        self.md_info_list = {}
+        if "packet_in" in ctrl_pkt_md:
+            self.p4_info = ctrl_pkt_md["packet_in"]
+            for md_info in self.p4_info.metadata:
+                self.md_info_list[md_info.name] = md_info
+        def _packet_in_recv_func(pkt_in):
+            global packet_in_handles
+            while True:
+                msg = client.get_stream_packet("packet", timeout=None)
+                if not msg:
+                    break
+
+                for _, handle in packet_in_handles.items():
+                    try:
+                        handle(msg.packet)
+                    except Exception as e:
+                        print(e)
+
+        self.recv_t = Thread(target=_packet_in_recv_func, args=(self, ))
+        self.recv_t.start()
+
+    def register_handler(self, handle, payload_regex=b'', **kwargs):
+        # Register a packet-in handle, will return the handle ID.
+        # Use keyword argument as additional filter to filter the packet in
+        # For example:
+        # register_handler(my_handle, payload_regex=b'.*\x02', egress_port='1')
+        global latest_hid
+        global packet_in_handles
+        filters = {}  # metadata id -> filter
+        if not handle or '__call__' not in dir(handle):
+            raise UserError("handle is not a callable object")
+
+        for key, value in kwargs.items():
+            if key not in self.md_info_list:
+                raise UserError("'{}' is not a packet-in metadata".format(key))
+            md_info = self.md_info_list[key]
+            filters[md_info.id] = bytes_utils.parse_value(value.strip(),
+                                                          md_info.bitwidth)
+
+        def _handle_wraper(packet_in_msg):
+            for md in packet_in_msg.metadata:
+                if md.metadata_id in filters:
+                    if md.value != filters[md.metadata_id]:
+                        return
+            if not re.search(payload_regex, packet_in_msg.payload):
+                return
+            handle(packet_in_msg)
+
+        latest_hid += 1
+        hid = latest_hid
+        packet_in_handles[hid] = _handle_wraper
+        return hid
```

Comment:
nit: in addition to returning the handler id, maybe we could add an attribute to the handler function object (`handler.hid = <hid>`) to make unregistering the handler easier